core/src/connection/manager.rs

```diff
@@ -0,0 +1,479 @@
+                }
+                Poll::Pending => return Poll::Pending,
+                Poll::Ready(None) => unreachable!("Manager holds both sender and receiver."),
+            }
+        };
+
+        if let hash_map::Entry::Occupied(mut task) = self.tasks.entry(*event.id()) {
+            Poll::Ready(match event {
+                task::Event::Notify { id: _, event } =>
+                    Event::ConnectionEvent {
+                        entry: EstablishedEntry { task },
+                        event
+                    },
+                task::Event::Established { id: _, info } => { // (2)
+                    task.get_mut().state = TaskState::Established(info); // (3)
+                    Event::ConnectionEstablished {
+                        entry: EstablishedEntry { task },
+                    }
+                }
+                task::Event::Error { id, error, handler } => {
+                    let id = ConnectionId(id);
+                    let task = task.remove();
+                    match handler {
+                        Some(handler) =>
+                            Event::PendingConnectionError { id, error, handler },
+                        None => match task.state {
+                            TaskState::Established(connected) =>
+                                Event::ConnectionError { id, connected, error },
+                            TaskState::Pending => unreachable!(
+                                "handler = `None` implies (2) occurred on that task and thus (3)."
+                            ),
+                        }
+                    }
+                }
+
+            })
+        } else {
+            unreachable!("By (1)")
+        }
+    }
+}
+
+/// An entry for a connection in the manager.
+#[derive(Debug)]
+pub enum Entry<'a, I, C> {
+    Pending(PendingEntry<'a, I, C>),
+    Established(EstablishedEntry<'a, I, C>)
+}
+
+impl<'a, I, C> Entry<'a, I, C> {
+    fn new(task: hash_map::OccupiedEntry<'a, TaskId, TaskInfo<I, C>>) -> Self {
+        match &task.get().state {
+            TaskState::Pending => Entry::Pending(PendingEntry { task }),
+            TaskState::Established(_) => Entry::Established(EstablishedEntry { task })
+        }
+    }
+}
+
+/// An entry for a managed connection that is considered established.
+#[derive(Debug)]
+pub struct EstablishedEntry<'a, I, C> {
+    task: hash_map::OccupiedEntry<'a, TaskId, TaskInfo<I, C>>,
+}
+
+/// An entry for a managed connection that is currently being established
+/// (i.e. pending).
+#[derive(Debug)]
+pub struct PendingEntry<'a, I, C> {
+    task: hash_map::OccupiedEntry<'a, TaskId, TaskInfo<I, C>>
+}
+
+impl<'a, I, C> PendingEntry<'a, I, C> {
+    /// Returns the connection id.
+    pub fn id(&self) -> ConnectionId {
+        ConnectionId(*self.task.key())
+    }
+
+    /// Aborts the pending connection attempt.
+    pub fn abort(self)  {
+        self.task.remove();
+    }
+}
+
+impl<'a, I, C> EstablishedEntry<'a, I, C> {
```

Comment:
Minor: Why not put this below `pub struct EstablishedEntry`? Why interleave `PendingEntry`?

Reply:
No reason, just the result of a lot of back-and-forth during the implementation, I guess.